fetch.bs

```diff
@@ -3241,7 +3253,111 @@ through TLS using ALPN. The protocol cannot be spoofed through HTTP requests in
 </ol>
 </div>
 
+<h3 id=https-upgrades>HTTPS upgrading</h3>
+
+<p>User agents may optionally upgrade requests with URLs that are not <a>potentially trustworthy
+URLs</a> to attempt to fetch them over <a>potentially trustworthy URLs</a>. If an upgraded request
+fails with a network error, it is retried over the original URL.
+
+<p>The HTTPS upgrading algorithm consists of <a>upgrade an HTTP request</a> and <a>HTTPS upgrade
+fallback</a> algorithms.
+
+<div algorithm>
+<h4 id=https-upgrades-upgrade>HTTPS upgrade algorithm</h4>
+
+<p>To <dfn>upgrade an HTTP request</dfn> given a <a for=/>request</a> <var>request</var>, run these
+steps:
+
+<ol>
+ <li>
+  <p>If one or more of the following conditions are met, return:
+  <ul>
+   <li><p><var>request</var>'s <a for="request">destination</a> is not "<code>document</code>"
+
+   <li><p><var>request</var>'s <a for="request">method</a> is not "<code>GET</code>"
 
+   <li><p><var>request</var>'s <a for="request">URL</a>'s <a for="url">scheme</a> is not
+   "<code>http</code>"
+
+   <li>
+    <p><var>request</var>'s <a for="request">URL</a>'s <a for="url">host</a> is exempted from
+    upgrades in an <a>implementation-defined</a> way.
+
+    <p class=example id="example-https-upgrades-exempted-hosts">If <a for=url>host</a> is a
+    non-registrable or non-assignable domain name such as .local or an IP address that falls in a
+    range reserved for non-publicly routable networks, the implementation might return without
+    modifying <var>request</var>.
```

Comment:
Do we really need to leave this implementation-defined? I guess we might have to for some things, but a lot of this is defined.

Reply:
Originally it was spelled out, but there didn't seem to be a spec that we could point to, so we changed the wording (also to allow more carveouts): https://github.com/whatwg/fetch/pull/1655#discussion_r1206670158

Do you prefer adding the "non-unique" wording from that discussion back in?

Reply:
I think my intuition would be more that we try to define the concept, as Mike West suggested. I would be okay with a follow-up issue that we highlight in the specification. (There should be some class=XXX already of that nature which you could imitate.)

Reply:
Just to be clear, do you mean something like this:
1. Define non-unique hostnames somewhere in this spec
2. Link to the definition from here
3. Add a note to the definition to point out that it should be moved to the Url spec

Reply:
One of these:

1. We admit this is underdefined for now and point to an open issue in an XXX marker.
2. We know the definition and therefore writing a PR against URL to add it there is not too much trouble.

Reply:
We discussed this more, and decided to remove the exemption for non-unique hostnames from upgrades in Chrome. There are a few reasons for this:
- The original reason for excluding these hostnames was that they were unlikely to serve valid HTTPS. However, this isn't universally true: developers may want to test and debug on environments where short names can serve valid HTTPS. We want to enable this use case.
- Browsers may (and most likely should) implement an allowlisting scheme where they don't try upgrades on a hostname if an upgrade failed on that hostname to avoid load time regressions (this is made possible by the "implementation-defined" wording above). In this scenario, an upgrade attempt on a non-unique hostname would only be done once every N days (current N=7 in Chrome, may change soon). So enabling HTTPS-Upgrades on non-unique hostnames should not cause a significant performance regression.
- As a side benefit, removing this carveout will make writing web platform tests for HTTPS-Upgrading possible. When run locally, WPTs are served on a non-unique hostname so HTTPS-Upgrades don't apply.

So I'm removing this example, but please let me know if you feel strongly otherwise.